jib-plugins-common/src/main/java/com/google/cloud/tools/jib/plugins/common/UpdateChecker.java

```diff
@@ -0,0 +1,193 @@
+            return Optional.empty();
+          }
+        } catch (DateTimeParseException ex) {
+          // Ignore parse failure; assume update check hasn't been performed
+        }
+      }
+
+      // Check for update
+      HttpURLConnection connection = (HttpURLConnection) new URL(versionUrl).openConnection();
+      try {
+        connection.setConnectTimeout(3000);
+        BufferedReader bufferedReader =
+            new BufferedReader(
+                new InputStreamReader(connection.getInputStream(), StandardCharsets.UTF_8));
+        String latestVersion = bufferedReader.readLine().trim();
+        Files.write(lastUpdateCheck, Instant.now().toString().getBytes(StandardCharsets.UTF_8));
+        if (currentVersion.equals(latestVersion)) {
+          return Optional.empty();
+        }
+        return Optional.of(
+            "A new version of Jib ("
+                + latestVersion
+                + ") is available (currently using "
+                + currentVersion
+                + "). Update your build configuration to use the latest features and fixes!");
+
+      } finally {
+        connection.disconnect();
+      }
+
+    } catch (IOException ex) {
+      try {
+        Files.deleteIfExists(configFile);
+        Files.deleteIfExists(lastUpdateCheck);
+      } catch (IOException ignored) {
+        // Fail silently
+      }
+    }
+
+    return Optional.empty();
+  }
+
+  /**
+   * Returns the config directory set by {@link PropertyNames#CONFIG_DIRECTORY} if not null,
+   * otherwise returns the default config directory.
+   *
+   * @return the config directory set by {@link PropertyNames#CONFIG_DIRECTORY} if not null,
+   *     otherwise returns the default config directory.
+   */
+  private static Path getConfigDir() {
+    String configDirProperty = System.getProperty(PropertyNames.CONFIG_DIRECTORY);
+    if (!Strings.isNullOrEmpty(configDirProperty)) {
+      return Paths.get(configDirProperty);
+    }
+    return XdgDirectories.getConfigHome();
+  }
+
+  private final Future<Optional<String>> updateMessageFuture;
+
+  @VisibleForTesting
+  UpdateChecker(Future<Optional<String>> updateMessageFuture) {
+    this.updateMessageFuture = updateMessageFuture;
+  }
+
+  /**
+   * Returns a message indicating Jib should be upgraded if the check succeeded and the current
+   * version is outdated, or returns {@code Optional.empty()} if the check was interrupted or did
+   * not determine that a later version was available.
+   *
+   * @return the {@link Optional} message to upgrade Jib if a later version was found, else {@code
+   *     Optional.empty()}.
+   */
+  public Optional<String> finishUpdateCheck() {
+    if (updateMessageFuture.isDone()) {
+      try {
+        return updateMessageFuture.get();
+      } catch (InterruptedException | ExecutionException ignored) {
+        // Fail silently;
+      }
+    }
+    updateMessageFuture.cancel(true);
```

Comment:
I think that this just tries to interrupt the process running the thread if it's still running? Do we need to check in `performUpdateCheck` that the thread hasn't been interrupted?

Reply:
I regarded this as an optional thing which may have virtually no effect in many cases. @TadCordle tested and observed that the build terminates right away when "the thread is still alive" (https://github.com/GoogleContainerTools/jib/pull/2214#discussion_r366007061). However, I just realized that the test using a single `Thread.sleep()` to simulate a  long-running thread is flawed, since I believe `Thread.sleep()` will unblock when a future is cancelled. @TadCordle the code to test this should look like this:

```java
while (true) {
  try {
    Thread.sleep(999999);
  } catch (Exception e) {}
}
```

Reply:
Still seems to terminate the thread immediately with the new test code.

Reply:
That's good. For both Maven and Gradle?

Reply:
Yup :+1: